lib/task-graph.js

```diff
@@ -61,13 +61,57 @@ function taskGraphFactory(
         };
         if (this.context.target) {
             this.logContext.id = this.context.target;
-        }
+       }
         // For database ref linking
         this.node = this.context.target || this.definition.options.nodeId;
 
         return this;
     }
 
+    /**
+    This is a helper that visits one node. 
+    This function is used in _visitGraphNode().
+    */
+
+    TaskGraph.prototype._visitOneGraphNode = function(task, value, dep) {
+        var nonTerminalOnStates = [];
+
+        // This won't actually get called because we do this same check in
+        // _populateTaskData, but keep this here just to be defensive.
+        if (!_.has(this.tasks, dep)) {
+            throw new Error('Graph does not contain task with ID ' + dep);
+        }
+        // Expand 'finished' to all finished task states
+        if (_.contains(value, Constants.Task.States.Finished)) {
+            value = Constants.Task.FinishedStates;
+            task.waitingOn[dep] = [Constants.Task.States.Finished];
+        }
+        // The dependent task is no longer terminal on state <value> since the
+        // current task lists that condition as a dependency.
+        // task.nonTerminalOnStates gets substracted from task.terminalOnStates
+        // when we traverse to the dependent task.
+        nonTerminalOnStates = Array.isArray(value) ? value : [value];
+
+        return nonTerminalOnStates;
+    };
+
+    /**
+    This is a helper that helps process a dependent task under waitOn key.
+    this function is used at _populateTaskData()
+    */
+    TaskGraph.prototype._processOneWaitOnTask = function(waitOnBase, idMap, waitOnTask) {
```

Comment:
I agree w/ @yyscamper; it's a little confusing especially because of the _visitGraphNode recursion. Something to the effect of 'do-stuff-with / get NonterminalStates'.